core/src/main/scala/kafka/server/DynamicBrokerConfig.scala

```diff
@@ -0,0 +1,378 @@
+  * (the "License"); you may not use this file except in compliance with
+  * the License.  You may obtain a copy of the License at
+  *
+  * http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+package kafka.server
+
+import java.nio.charset.StandardCharsets
+import java.util
+import java.util.Properties
+import java.util.concurrent.locks.ReentrantReadWriteLock
+
+import kafka.server.DynamicBrokerConfig._
+import kafka.utils.{CoreUtils, Logging}
+import kafka.zk.{AdminZkClient, KafkaZkClient}
+import org.apache.kafka.common.Reconfigurable
+import org.apache.kafka.common.config.types.Password
+import org.apache.kafka.common.config.{ConfigDef, ConfigException, SslConfigs}
+import org.apache.kafka.common.network.ListenerReconfigurable
+import org.apache.kafka.common.utils.Base64
+
+import scala.collection._
+import scala.collection.JavaConverters._
+
+/**
+  * Dynamic broker configurations are stored in ZooKeeper and may be defined at two levels:
+  * <ul>
+  *   <li>Per-broker configs persisted at <tt>/configs/brokers/brokerId</tt>: These can be described/altered
+  *       using AdminClient using the resource name brokerId.</li>
+  *   <li>Cluster-wide defaults persisted at <tt>/configs/brokers/&lt;default&gt;</tt>: These can be described/altered
+  *       using AdminClient using an empty resource name.</li>
+  * </ul>
+  * The order of precedence for broker configs is:
+  * <ol>
+  *   <li>DYNAMIC_BROKER_CONFIG: stored in ZK at /configs/brokers/brokerId</li>
+  *   <li>DYNAMIC_DEFAULT_BROKER_CONFIG: stored in ZK at /configs/brokers/&lt;default&gt;</li>
+  *   <li>STATIC_BROKER_CONFIG: properties that broker is started up with, typically from server.properties file</li>
+  *   <li>DEFAULT_CONFIG: Default configs defined in KafkaConfig</li>
+  * </ol>
+  * Log configs use topic config overrides if defined and fallback to broker defaults using the order of precedence above.
+  * Topic config overrides may use a different config name from the default broker config.
+  * See [[kafka.log.LogConfig#TopicConfigSynonyms]] for the mapping.
+  * <p>
+  * AdminClient returns all config synonyms in the order of precedence when configs are described with
+  * <code>includeSynonyms</code>. In addition to configs that may be defined with the same name at different levels,
+  * some configs have additional synonyms.
+  * </p>
+  * <ul>
+  *   <li>Listener configs may be defined using the prefix <tt>listener.name.listenerName.configName</tt>. These may be
+  *       configured as dynamic or static broker configs. Listener configs have higher precedence than the base configs
```

Comment:
To clarify my understanding, you must override listener configs at the broker level? So if all brokers share the same listener, and you want to update a listener property, you have to do it for every broker?

Reply:
Yes, in the current implementation, listener configs can be configured only per-broker. Some listener configs have to be defined at the broker level anyway (e.g. keystore/keystore password, Kerberos JAAS config that contains hostname). And the actual listeners/advertised listeners also need to be per-broker since they may contain host. At least initially, I thought it would best to keep all these related configs together so that they get reconfigured at the same time. Keystores are the only configs for which we allow updates in this KIP and these need to be per-broker. For other configs, we only support addition of new listeners, so configuring per-broker with all the listener configs should be ok?